x/smart/stream_dialer.go

```diff
@@ -292,23 +309,60 @@ func (f *StrategyFinder) findTLS(ctx context.Context, testDomains []string, base
 	}), nil
 }
 
-// NewDialer uses the config in configBytes to search for a strategy that unblocks DNS and TLS for all of the testDomains, returning a dialer with the found strategy.
-// It returns an error if no strategy was found that unblocks the testDomains.
-// The testDomains must be domains with a TLS service running on port 443.
-func (f *StrategyFinder) NewDialer(ctx context.Context, testDomains []string, configBytes []byte) (transport.StreamDialer, error) {
-	var parsedConfig configConfig
-	err := yaml.Unmarshal(configBytes, &parsedConfig)
-	if err != nil {
-		return nil, fmt.Errorf("failed to parse config: %v", err)
+// Return the fastest fallback dialer that is able to access all the testDomans
+func (f *StrategyFinder) findFallback(ctx context.Context, testDomains []string, fallbackConfig []string) (transport.StreamDialer, error) {
+	if len(fallbackConfig) == 0 {
+		return nil, errors.New("no fallback was specified")
 	}
 
-	// Make domain fully-qualified to prevent confusing domain search.
-	testDomains = append(make([]string, 0, len(testDomains)), testDomains...)
-	for di, domain := range testDomains {
-		testDomains[di] = makeFullyQualified(domain)
+	ctx, searchDone := context.WithCancel(ctx)
+	defer searchDone()
+	raceStart := time.Now()
+	type SearchResult struct {
+		Dialer transport.StreamDialer
+		Config string
+	}
+
+	fallback, err := raceTests(ctx, 250*time.Millisecond, fallbackConfig, func(fallbackUrl string) (*SearchResult, error) {
+		dialer, err := getStreamDialer(ctx, fallbackUrl)
+		if err != nil {
+			return nil, fmt.Errorf("getStreamDialer failed: %w", err)
+		}
+
+		for _, testDomain := range testDomains {
```

Comment:
If you'd like to clean it up, this for-loop can be factored out, since it's the same as the one for TLS. Perhaps call it something like `testDialer(dialer, domains)`

Reply:
refactored out